external/fv3fit/fv3fit/data/tfdataset.py

```diff
@@ -0,0 +1,43 @@
+import dataclasses
+from typing import Mapping, Sequence, Optional
+import tensorflow as tf
+from .base import TFDatasetLoader
+
+
+@dataclasses.dataclass
+class VariableConfig:
+    unstacked_dims: Sequence[str]
+
+
+@dataclasses.dataclass
+class WindowedZarrLoader(TFDatasetLoader):
+    """
+    A tfdataset loader that loads directly from zarr and supports time windows.
+
+    Attributes:
+        data_path: path to zarr data
+        window_size: number of timesteps to include in time windows
+        default_variable_config: default configuration for variables
+        variable_configs: configuration for variables by name
+    """
+
+    data_path: str
+    window_size: int
+    default_variable_config: VariableConfig
+    variable_configs: Mapping[str, VariableConfig] = dataclasses.field(
+        default_factory=dict
+    )
+
+    def get_data(
+        self, local_download_path: Optional[str], variable_names: Sequence[str],
```

Comment:
I had to keep this as part of the API in order to avoid making breaking changes to the config. The local download path is currently part of a caching strategy object which is effectively part of the top-level training configuration, and not part of the data configuration. To move local_download_path into e.g. BatchesLoader, this would have to become part of the data yaml config, which is fairly inconvenient (we like being able to set it as a command-line argument).

I'd like to not do it in this PR, but I have an idea on how to do this in `load_data` as you suggested that I could implement after discussing it in a standup or on slack. If I switch the usage of local_download_path to do caching at the tfdataset level (`tfdataset.cache(local_download_path)`) and apply it after the get_data call. This would solve the issue, but means the locally cached data will be saved in the form of tfrecords, not as netcdf files. This means if anyone was manually using the cache directory, it might break their workflow.

I don't think this will be an issue - the suggested way to make local cache directories to later load from is to use the save functionality from loaders directly - but I've used this before in some local code so I want to make sure everyone is aware of this change and not depending on the format of the local cache directory.

Reply:
> If I switch the usage of local_download_path to do caching at the tfdataset level (tfdataset.cache(local_download_path)) and apply it after the get_data call. This would solve the issue, but means the locally cached data will be saved in the form of tfrecords, not as netcdf files. 

We had memory issues when using `.cache` like this sadly. Also, `.cache` doesn't save tfrecords. It's some undocumented format. I spent a while debugging it a few months ago and felt it was easier to make a leak-free caching by rolling our own caching layer.

Reply:
Yeah, rolling our own caching does seem simpler than debugging a memory issue.